substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/sampler/SamplerJfrStackTraceSerializer.java

```diff
@@ -167,8 +169,12 @@ private static int visitFrame(JfrNativeEventWriterData data, CodeInfo codeInfo,
         int numStackTraceElements = 0;
         FRAME_INFO_CURSOR.initialize(codeInfo, ip, false);
         while (FRAME_INFO_CURSOR.advance()) {
+            FrameInfoQueryResult frame = FRAME_INFO_CURSOR.get();
+            if (SubstrateJVM.getStackTraceRepo().getTrimInternalStackTraces() &&
+                            UninterruptibleUtils.String.startsWith(frame.getSourceClassName(), SUBSTRATEVM_PREFIX)) {
+                continue;
+            }
             if (data.isNonNull()) {
-                FrameInfoQueryResult frame = FRAME_INFO_CURSOR.get();
                 serializeStackTraceElement(data, frame);
```

Comment:
Should this additional check be within the `if (data.isNonNull()) {` branch or not?

Reply:
Yes, maybe moving it inside `if (data.isNonNull())` would be a small optimization

Reply:
Actually, I don't think it should be moved inside the `if (data.isNonNull())` check. We must increment `numStackTraceElements` regardless of the if (data.isNonNull())` check, but must not increment if we are trimming the current stacktrace.